### ISSUE TITLE
add all subkey fingerprints, and uid emails, into KeyInfo as optional arrays

### ISSUE THREAD

**Issue by tomholub** [FlowCrypt/flowcrypt-browser#1640]
The `KeyInfo` format is missing information about subkey longids / fingerprints, and also about emails listed in the key.

If we had this information in storage, we could avoid parsing keys when we need only the subkey longids, eg:
```js
      for (const ki of kiWithPp.keys) {
        // this is inefficient because we are doing unnecessary parsing of all keys here
        // better would be to compare to already stored KeyInfo, however KeyInfo currently only holds primary longid, not longids of subkeys
        // while messages are typically encrypted for subkeys, thus we have to parse the key to get the info
        // we are filtering here to avoid a significant performance issue of having to attempt decrypting with all keys simultaneously
        const { ids } = await Pgp.key.details(await Pgp.key.read(ki.private));
        for (const { longid } of ids) {
          if (keys.encryptedFor.includes(longid)) {
            keys.prvMatching.push(ki);
            break;
          }
        }
      }
```
These new fields can be added as optional type. We always add it when creating KeyInfo, plus we'll add a migration in background_page that adds these for existing keys.

For example:

```diff
export interface PrvKeyInfo {
  private: string;
  longid: string;
+ fingerprints?: string[];
+ emails?: string[];
  passphrase?: string;
  decrypted?: Key;  // only for internal use in this file
  parsed?: Key;     // only for internal use in this file
}
```

Then in a future release, in another PR a few months later, when we're sure it went through the migration and works without issues, we can make the `fingerprints` and `emails` no longer optional, completely remove the `longid` filed from `KeyInfo`.

**Comment by tomholub** [comment]
also useful for performance of #2154, #1849, #670

**Comment by tomholub** [comment]
would be more pleasant after #2520 landed

**Comment by rrrooommmaaa** [comment]
@tomholub to convert all entries in the storage we should add a method like getAll to `AbstractStore`, right?

**Comment by tomholub** [comment]
You can do it with standard chrome APIs, in particular `storageLocalGetAll`. But actually you may be able to use `storageLocalGet` if you give it `keys` as the storage index. You can experiment and see.

Generally, I don't add methods that are used only for migration into the regular classes. Migration code is expected to be added, then removed a few months later, so I like it to be self-contained, ideally not polluting existing code in more than one place.

**Comment by tomholub** [comment]
Actually, even differently: you can, in migration code, retrieve all account emails using `GlobalStore.acctEmailsGet`. Then you do `KeyStore.get` without specifying fingerprints. Then you do `KeyStore.add` for each of the keys back, making sure to add the needed properties. Finally, you could add something like `localStore.key_info_store_fingerprints_added = true` and check for that - if it's already true, you skip migration next time. Then eventually we delete that code.

**Comment by rrrooommmaaa** [comment]
There is also parsing in SettingsView.addKeyRowsHtml to get `created` value. Leave it as it is or add `created` to `KeyInfo`?

**Comment by tomholub** [comment]
Good question, sure - could add it as well.

**Comment by tomholub** [comment]
I was actually thinking to add whole `Key` object instead of `KeyInfo` and stop using `KeyInfo`. But I was worried maybe the change is too large. What do you think?

It would have to be a superset of `Key`, beause it still needs the pass phrase, armored private key, armored public key, ...

**Comment by tomholub** [comment]
That would look something like this:
```
{
   armoredPrv: string,
   armoredPub: string,
   key: Key,
   passphrase: string?
}
```
But the problem with this approach is that if we change definition of `Key` like you recently did, it will look like the storage has updated schema but in fact it doesn't. So that's not good either.

Probably ok to leave existing structure, just add a few fields.

**Comment by rrrooommmaaa** [comment]
Interestingly, we already have `Key` (`pubkey` field) in Contact store.
Would the latest PR make it unusable?

**Comment by tomholub** [comment]
I don't think that actually gets stored into storage. Can you check by
pulling it from storage?

I think the TS types are getting conflated, maybe the code needs to be
updated to be clearer.

**Comment by rrrooommmaaa** [comment]
Of course, it's in the storage.
![image](https://user-images.githubusercontent.com/23335944/100904590-37485680-34d8-11eb-8cb1-5022e3920d42.png)

**Comment by tomholub** [comment]
I think you're looking at contacts (public keys). They use IDBDatabase.

`key-store.ts` stores keys using chrome extension storage - have a look at that one?

Unless I saw wrong.

**Comment by rrrooommmaaa** [comment]
contacts are in IndexedDB, yes. My concern is that we don't use `usableButExpired` flag anymore (we renamed it to usableForEncryptionButExpired). Don't we need to do migration update for this?

**Comment by tomholub** [comment]
Well noticed, this could lead to endless bugs.

Migration may be difficult because some users may have many contacts. I have easily 5000 - it would take minutes to load and parse. It's possible - do it slowly one by one with some delay in between, in background process. But I'd like to see if an easier solution may be possible.

I think our best bet would be to instead untangle the _type_ of that stored object from `Key` type. We can just describe it as a separate type, like `StoredContactKey` or whatever. Then, if someone tries to access a property that's not on it, we'll see a compiler error.

When we change `Key`, we don't change `StoredContactKey`. If a consumer needs access to a field that's on `Key` but not on `StoredContactKey`, they'll have to parse the pubkey after retrieving it from the db. If we start doing a lot of that, then we can consider a migration of contacts db too.

**Comment by tomholub** [comment]
Also this means, before storing, we have to manually construct the `StoredContactKey` object instead of just lazily declaring `key as StoredContactKey` because that of course wouldn't help. We need to control exactly what gets stored and what comes out.

I have not released your recent changes, so no actual customer browser storage were "polluted" with the changed Key structure yet.

**Comment by tomholub** [comment]
Maybe you'll have some other idea

**Comment by rrrooommmaaa** [comment]
Sure, the database entities (StoreContactKey) should be extended/transformed to `Key` upon loading.
Some properties (like Usable, Expired ...) should not be stored because they depend on current date

**Comment by tomholub** [comment]
>  Some properties (like Usable, Expired ...) should not be stored because they depend on current date

Good point

> Sure, the database entities (StoreContactKey) should be extended/transformed to Key upon loading.

Only if it doesn't involve parsing the armored key. Parsing armored key takes about 10ms per key, when you load many contacts, this becomes a problem. see https://github.com/FlowCrypt/flowcrypt-browser/issues/3155
